File: LeetCode/Graph/DFS/Golang/Leetcode473-matchsticks-to-square.go
package sbgo

import "sort"

func makesquare(arr []int) bool {
	sum := 0
	for _, val := range arr {
		sum += val
	}
	if sum%4 != 0 {
		return false
	}
	sort.Slice(arr, func(i, j int) bool {
		return arr[i] > arr[j]
	})
	edge := make([]int, 4)
	var dfs func(u int) bool
	dfs = func(u int) bool {
		if u == len(arr) {
			return true
		}
		for i := 0; i < 4; i++ {
			edge[i] += arr[u]
			if edge[i] <= sum/4 && dfs(u+1) {
				return true
			}
			edge[i] -= arr[u]
		}
		return false
	}
	return dfs(0)
}

File: LeetCode/Graph/DFS/Python/Leetcode473-matchsticks-to-square.py
class Solution:
    def makesquare(self, arr: List[int]) -> bool:
        tot = sum(arr)
        if tot % 4 != 0: return False
        arr.sort(reverse=True)
        edge = [0] * 4

        def dfs(u):
            if u == len(arr): return True
            for i in range(4):
                edge[i] += arr[u]
                if edge[i] <= tot // 4 and dfs(u + 1):
                    return True
                edge[i] -= arr[u]
            return False

        return dfs(0)
